rest/functionsapitest/user_functions_test.go

```diff
@@ -0,0 +1,1463 @@
+			},
+		},
+	}
+
+	n1qlBackTickTestCfgDoubleDollar := functions.FunctionsConfig{
+		Definitions: functions.FunctionsDefs{
+			"user": &functions.FunctionConfig{
+				Type:  "query",
+				Code:  "SELECT $$user AS `user`", // use backticks for n1ql reserved keywords
+				Allow: allowAll,
+			},
+		},
+	}
+
+	rt := rest.NewRestTesterForUserQueries(t, rest.DbConfig{})
+	if rt == nil {
+		return
+	}
+	defer rt.Close()
+
+	t.Run("using admin end point", func(t *testing.T) {
+
+		// positive cases:
+		t.Run("using single dollar", func(t *testing.T) {
+			request, err := json.Marshal(userFuncConfigSingleDollar)
+			assert.NoError(t, err)
+			response := rt.SendAdminRequest("PUT", "/db/_config/functions", string(request))
+			assert.Equal(t, 200, response.Result().StatusCode)
+		})
+
+		t.Run("using single dollar and backtick", func(t *testing.T) {
+			request, err := json.Marshal(n1qlBackTickTestCfgSingleDollar)
+			assert.NoError(t, err)
+			response := rt.SendAdminRequest("PUT", "/db/_config/functions", string(request))
+			assert.Equal(t, 200, response.Result().StatusCode)
+		})
+
+		// negative cases:
+		t.Run("using double dollar", func(t *testing.T) {
+			request, err := json.Marshal(userFuncConfigDoubleDollar)
+			assert.NoError(t, err)
+			response := rt.SendAdminRequest("PUT", "/db/_config/functions", string(request))
+			assert.Equal(t, 400, response.Result().StatusCode)
+			assert.Contains(t, string(response.BodyBytes()), "illegal 'allow' pattern")
+		})
+
+		t.Run("using double and backtick", func(t *testing.T) {
+			// configure query
+			request, err := json.Marshal(n1qlBackTickTestCfgDoubleDollar)
+			assert.NoError(t, err)
+			response := rt.SendAdminRequest("PUT", "/db/_config/functions", string(request))
+			assert.Equal(t, 200, response.Result().StatusCode)
+
+			// execute query
+			response = rt.SendAdminRequest("GET", "/db/_function/user", "")
+			assert.Equal(t, 500, response.Result().StatusCode)
+			assert.Contains(t, string(response.BodyBytes()), "Internal Server Error")
+			assert.Contains(t, string(response.BodyBytes()), "syntax error")
+		})
+	})
+
+	t.Run("using DB Config", func(t *testing.T) {
+		createDB := func(t *testing.T, funcCfg functions.FunctionsConfig) (*rest.TestResponse, error) {
+			dbCfg := rest.GetBasicDbCfg(rt.TestBucket)
+			dbCfg.UserFunctions = &funcCfg
+
+			rt.SendAdminRequest(http.MethodDelete, "/db/", "")
+
+			resp, err := rt.CreateDatabase("db", dbCfg)
+			return resp, err
+		}
+
+		// positive cases:
+		t.Run("using double dollar", func(t *testing.T) {
+			resp, err := createDB(t, userFuncConfigDoubleDollar)
+			assert.NoError(t, err)
+			assert.Equal(t, 201, resp.Result().StatusCode)
+		})
+
+		// negative cases:
+		t.Run("using single dollar", func(t *testing.T) {
+			resp, err := createDB(t, userFuncConfigSingleDollar)
+			assert.NoError(t, err)
+			assert.Equal(t, 500, resp.Result().StatusCode)
+			assert.Contains(t, string(resp.BodyBytes()), "undefined environment variable")
```

Comment:
instead of an using an environment variable named `user` , 
what if I keep on generating a random variable until it is not present in the list of environment variables. 

Can we use that to test this ? 
I am asking this because, I feel like we should test if $$variables are properly parsed and passed to functions package as expected.